x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java

```diff
@@ -259,38 +262,68 @@ private void ensureOpen() {
         }
     }
 
-    CompletableFuture<Integer> fetchRange(
-        long start,
-        long end,
-        CheckedBiFunction<Long, Long, Integer, IOException> onRangeAvailable,
-        CheckedBiConsumer<Long, Long, IOException> onRangeMissing
+    @FunctionalInterface
+    interface CacheReader {
+        int read(FileChannel channel) throws IOException;
+    }
+
+    @FunctionalInterface
+    interface CacheWriter {
+        void write(FileChannel channel, long from, long to, Consumer<Long> progressUpdater) throws IOException;
+    }
+
+    CompletableFuture<Integer> fetchAsync(
+        final Tuple<Long, Long> rangeToWrite,
+        final Tuple<Long, Long> rangeToRead,
+        final CacheReader reader,
+        final CacheWriter writer,
+        final Executor executor
     ) {
         final CompletableFuture<Integer> future = new CompletableFuture<>();
         try {
-            if (start < 0 || start > tracker.getLength() || start > end || end > tracker.getLength()) {
-                throw new IllegalArgumentException(
-                    "Invalid range [start=" + start + ", end=" + end + "] for length [" + tracker.getLength() + ']'
-                );
-            }
             ensureOpen();
             final List<SparseFileTracker.Gap> gaps = tracker.waitForRange(
-                Tuple.tuple(start, end),
-                Tuple.tuple(start, end), // TODO use progressive sub range to trigger read operations sooner
-                ActionListener.wrap(
-                    rangeReady -> future.complete(onRangeAvailable.apply(start, end)),
-                    rangeFailure -> future.completeExceptionally(rangeFailure)
-                )
+                rangeToWrite,
+                rangeToRead,
+                ActionListener.wrap(success -> future.complete(reader.read(channel)), future::completeExceptionally)
```

Comment:
Can we assert that `reader.read(channel)` returns `rangeToRead.end() - rangeToRead.start()` here?

Reply:
Good suggestion, thanks